### PR TITLE
Update client monorepo to eslint 7

### PR DESCRIPTION
**Pull request by tylerbutler**
This PR upgrades the client monorepo to ESLint 7. During the course of doing this, I noted a few rules that are worth discussing more.

[@typescript-eslint/strict-boolean-expressions](https://github.com/typescript-eslint/typescript-eslint/blob/v4.2.0/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md)

This rules requires `strictNullChecks=true` in tsconfig in order to be useful. Many of our packages don't have this enabled (for various reasons). But they really should, no? Anyway, the docs indicate that the rule is useless without this config, so it's ignored in a lot of projects.

[@typescript-eslint/restrict-template-expressions](https://github.com/typescript-eslint/typescript-eslint/blob/v4.2.0/packages/eslint-plugin/docs/rules/restrict-template-expressions.md)

We violate this a lot, but I didn't notice anything terrible in the cases I spot checked. My recommendation is we leave this disabled.

[@typescript-eslint/no-unsafe-return](https://github.com/typescript-eslint/typescript-eslint/blob/v4.2.0/packages/eslint-plugin/docs/rules/no-unsafe-return.md)

We could improve type-safety by enabling this rule and fixing what it finds. But the violations are _numerous..._

[@typescript-eslint/ban-types](https://github.com/typescript-eslint/typescript-eslint/blob/v4.2.0/packages/eslint-plugin/docs/rules/ban-types.md)

We use the `object`, `{}`, and `Function` types quite a bit, so we aren't getting as much type safety as we could be. From the docs:

-   Avoid the Function type, as it provides little safety for the following reasons:
    -   It provides no type safety when calling the value, which means it's easy to provide the wrong arguments.
    -   It accepts class declarations, which will fail when called, as they are called without the new keyword.

-   Avoid the Object and {} types, as they mean "any non-nullish value".
    -   This is a point of confusion for many developers, who think it means "any object type".

-   Avoid the object type, as it is currently hard to use due to not being able to assert that keys exist.
    -   See [microsoft/TypeScript#21732](https://github.com/microsoft/TypeScript/issues/21732).

These _do_ seem worth fixing at some point, but I'm not sure what we should replace `Object` with.

### CODE REVIEW

File: examples/data-objects/webflow/src/clipboard/paste.ts [hunk 1]
```diff
@@ -21,7 +21,8 @@ export function paste(doc: FlowDocument, data: DataTransfer, position: number) {
         const root = document.createElement("span");
         root.innerHTML = content;
         pasteChildren(doc, root, position);
-    } else if (content = data.getData(ClipboardFormat.html)) {
+        // ESLint found a bug!
+    } else if (content = data.getData(ClipboardFormat.text)) {
```
Comment (tylerbutler) [inline examples/data-objects/webflow/src/clipboard/paste.ts:25]: I think ESLint found a bug here, and I made a fix. Is it the right one?

**Comment by tylerbutler** [comment]
I recognize it's difficult to review so many changes in a single PR, so I am open to suggestions for how to make this easier to review without breaking the monorepo build.

File: examples/data-objects/webflow/src/clipboard/paste.ts (hunk 1 above)
Comment (markfields) [reply examples/data-objects/webflow/src/clipboard/paste.ts:25]: Can you avoid making the bug fix in this PR, so it's isolated in history?

Comment (tylerbutler) [reply examples/data-objects/webflow/src/clipboard/paste.ts:25]: Yes, good call. I'll revert this.

File: examples/data-objects/shared-text/src/pageCacher.ts [hunk 2]
```diff
@@ -11,7 +11,9 @@ interface IWindow extends Window {
 // Extract the head and body HTML. Remove any <script> tag. And then wrap inside a basic HTML page.
 export function createCacheHTML(): void {
     const [bodyHTML, headHTML] = [document.body.innerHTML, document.head.innerHTML];
+    // eslint-disable-next-line unicorn/no-unsafe-regex
```
Comment (markfields) [inline examples/data-objects/shared-text/src/pageCacher.ts:14]: This seems scary - potentially exponential-time, according to this?  https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-unsafe-regex.md

File: packages/runtime/runtime-utils/src/serializer.ts [hunk 3]
```diff
@@ -37,7 +37,8 @@ export class FluidSerializer implements IFluidSerializer {
             // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
             const handle = !!value && value.IFluidHandle;
             // TODO - understand why handle === false in some of our tests
-            // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+            // eslint-disable-next-line max-len
```
Comment (markfields) [inline packages/runtime/runtime-utils/src/serializer.ts:40]: lol

File: packages/test/end-to-end-tests/src/test/detachedContainerTests.spec.ts [hunk 4]
```diff
@@ -336,7 +333,6 @@ const tests = (args: ICompatTestArgs) => {
         const router = await dataStore.context.containerRuntime.createDataStore([testDataStoreType]);
         const comp2 = await requestFluidObject<ITestFluidObject>(router, "/");
 
-        // eslint-disable-next-line @typescript-eslint/unbound-method
```
Comment (markfields) [inline packages/test/end-to-end-tests/src/test/detachedContainerTests.spec.ts:339]: What happened to this rule?

File: examples/data-objects/webflow/src/clipboard/paste.ts (hunk 1 above)
Comment (tylerbutler) [reply examples/data-objects/webflow/src/clipboard/paste.ts]: PR #3836.

**Review by markfields — APPROVED** [review]
🎂 I only reviewed the .ts and .tsx files. All looks pretty reasonable.

I did leave some comments, take a look. And to add to your list from the initial comment, I'm curious about the module-related ones, like no-var-requires etc. I don't know much about how that works and what goes bad with the different syntax options.

File: packages/test/end-to-end-tests/src/test/detachedContainerTests.spec.ts (hunk 4 above)
Comment (tylerbutler) [reply packages/test/end-to-end-tests/src/test/detachedContainerTests.spec.ts:339]: The rule is still enabled, so it was either a false positive before or a false negative now. Not sure which.

**Comment by markfields** [comment]
🎉

**Pull request merged**